### ISSUE TITLE
`CA1062` gives a false positive on tuple patterns

### ISSUE THREAD

**Issue by RenderMichael** [dotnet/roslyn-analyzers#6977]
### Analyzer

**Diagnostic ID**: [CA1062](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1062): `Validate arguments of public methods`

### Analyzer source

.NET 8 RC1

### Describe the bug

```csharp
public class C : System.IEquatable<C>
{
    public static bool operator ==(C? one, C? other) => (one, other) switch
    {
        (null, null) => true,
        (null, _) or (_, null) => false,
        _ => one.Equals(other) // CA1062: validate parameter `one` is not-null before using it
    };

    public static bool operator !=(C? one, C? other) => !(one == other);

    public bool Equals(C? other) => true;

    public override bool Equals(object? obj) => Equals(obj as C);

    public override int GetHashCode() => 0;
}
```

### Expected behavior

No Warning.

### Actual behavior

False Positive.

**Comment by mavasani** [comment]
The above test code is equivalent to the below, which also fires this diagnostic:

```csharp
if (one == null && other == null)
    return true;

if (one == null && other != null)
    return false;

return one.Equals(other);
```

To correctly identify that `one` is guaranteed to be non-null, we need to associatively track that all values of `other` for `one == null` have been accounted for prior to `one.Equals` dereference.
This issue is general shortcoming of any dataflow analysis as we cannot track relative values of multiple symbols on different code paths to avoid exploding the value tracking graph. This is by design, and you should add a source suppression or an explicit null check here for clarity and better readability of code.
